zest/releaser/vcs.py

```diff
@@ -204,33 +246,41 @@ def _update_version(self, version):
             setup_version = self.get_setup_py_version()
             if not setup_version or (setup_version ==
                                      self.get_version_txt_version()):
-                open(versionfile, 'w').write(version + '\n')
-                logger.info("Changed %s to %r", versionfile, version)
+                open(versionfile, 'w').write(version + u'\n')
+                logger.info(
+                    u"Changed {0} to {1!r}".format(versionfile, version)
+                    )
                 return
 
-        good_version = "version = '%s'" % version
-        line_number = 0
-        setup_lines = open('setup.py').read().split('\n')
-        for line_number, line in enumerate(setup_lines):
-            match = VERSION_PATTERN.search(line)
-            if match:
-                logger.debug("Matching version line found: %r", line)
-                if line.startswith(' '):
-                    # oh, probably '    version = 1.0,' line.
-                    indentation = line.split('version')[0]
-                    # Note: no spaces around the '='.
-                    good_version = indentation + "version='%s'," % version
-                setup_lines[line_number] = good_version
-                contents = '\n'.join(setup_lines)
-                open('setup.py', 'w').write(contents)
-                logger.info("Set setup.py's version to %r", version)
-                return
+        content = read_text_file(filename)
+        tree = ast.parse(content, filename)
+        assignment = self._find_assignment(tree, u'version')
+        if assignment:
+            # This assumes that the version string does not span more than
+            # one line
+            position = (
+                assignment.value.lineno - 1,
+                assignment.value.col_offset,
+                assignment.value.col_offset + len(repr(assignment.value.s))
+                )
+            lines = content.splitlines()
+            logger.debug(u"Matching version line found: {0!r}".format(
+                lines[position[0]]))
+            lines[position[0]] = (
+                lines[position[0]][:position[1]] +
+                repr(version) +
+                lines[position[0]][position[2]:]
+                )
+            contents = u'\n'.join(lines)
+            open('setup.py', 'w').write(contents)
+            logger.info("Set setup.py's version to {0}".format(version))
+            return
```

Comment:
In the find_assignment method it also looks for "keyword"  which behaves
similarly.
On 2 Jun 2015 7:58 am, "Marius Gedminas" <notifications@github.com> wrote:

> In zest/releaser/vcs.py
> <https://github.com/zestsoftware/zest.releaser/pull/97#discussion_r31496745>
> :
>
> > +                assignment.value.lineno - 1,
> > +                assignment.value.col_offset,
> > +                assignment.value.col_offset + len(repr(assignment.value.s))
> > +                )
> > +            lines = content.splitlines()
> > +            logger.debug(u"Matching version line found: {0!r}".format(
> > +                lines[position[0]]))
> > +            lines[position[0]] = (
> > +                lines[position[0]][:position[1]] +
> > +                repr(version) +
> > +                lines[position[0]][position[2]:]
> > +                )
> > +            contents = u'\n'.join(lines)
> > +            open('setup.py', 'w').write(contents)
> > +            logger.info("Set setup.py's version to {0}".format(version))
> > +            return
>
> I don't see where you're handling
>
> setup(
>     ...
>     version='1.2.3.dev0',
>     ...
> )
>
> which is not an assignment.
>
> The old code used to handle that (that's what the # oh, probably '
> version = 1.0,' line. comment was about).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zestsoftware/zest.releaser/pull/97/files#r31496745>.
>